Make the Venafi Issuer handle oauth refresh tokens

wallrj opened the pull request:
* Refactored venaficlient package to optionally support getting new access-token in TPP, using a refresh-token
* Moved Venafi issuer and clusterissuer handling to a standalone controller (to allow me to access the queue)
* Request new access-token if the current access-token causes authentication errors and is beyond it's recorded expiry time
* requeue the item for  re-reconciliation every minute to ensure that the token is up-to-date. 

TODO:
 - [ ] add unit tests venaficlient package
 - [ ] improve rescheduling so that we always reschedule before the access-token expires.

**Release note**:
```release-note
Venafi Issuer now supports oauth refresh tokens
```

munnerz commented on pkg/issuer/venafi/client/venaficlient.go:
```diff
@@ -70,15 +82,90 @@ type connector interface {
+				return nil, fmt.Errorf("error creating patch: %v", err)
+			}
+
+			_, err = coreClient.Secrets(namespace).Patch(
```
Is there a reason for using a Patch here? Do we anticipate any other actors concurrently modifying this Secret resource?

munnerz commented on pkg/issuer/venafi/client/venaficlient.go:
```diff
@@ -70,15 +82,90 @@ type connector interface {
+		return venCfg.TPP.CredentialsRef.Name
+	case venCfg.Cloud != nil:
+		return venCfg.Cloud.APITokenSecretRef.Name
+	}
```
So just to confirm, the new refresh token is being written into the same user-supplied Secret resource? What happens if someone is using a GitOps like controller to create that Secret in the first place? Won't the GitOps controller 'delete' the newly written refresh token, meaning the entire auth flow has to be re-done due to the refresh token being lost?

munnerz reviewed (commented):
I voiced this on the stand up call that I don't think cert-manager should be persisting the refresh token and relying on it as the means to authenticate. Given this is specific to the Venafi issuer, I don't want to 'hold' this back if you feel it is in fact okay.

The problem areas I see here:

1) writing back into the same user-provided Secret resource: this is something that has caused us a load of problems in that past with the Certificate resources `spec.secretName` - we conflated the 'input' (private key) and the output (certificate) into a single resource. In a similar manner, we are conflating the input, and the output (the refresh token)

2) refresh tokens are not typically persisted this way. In a usual oauth flow, some other credential is exchanged for an access token and refresh token. This allows for the client to re-authenticate in the event of the refresh token expiring. Making the refresh token valid for extremely long periods of time to avoid this undermines the integrity of oauth2 altogether, and you'd be better to avoid all this complexity and simply have a long lived access token (or username/password pair). They are functionally equivalent, and mixing oauth into this is introducing unnecessary complexity.

3) the venafi client stuff here was intended to be a read-only cache, where the cache is populated by a controller that watches for venafi-typed issuer resources. This creates a clear separation between what constructs Venafi clients, and what utilises them. This change now puts a *write* operation onto the *read* path. If you are going to do this, I'd strongly recommend adding the logic for persisting and rotating refresh tokens into a distinct controller and keeping the *read* path as simple as possible (i.e. looking up a 'venafi client' in a map of issuer->client structure)

Overall, I'm quite strongly against this change (as we've already discussed). I think it reduces reliability, increases complexity, and has a net *negative* on the security of the system.

Given this is ultimately a vendor-specific change however, I wouldn't want to put an executive 'hold' on it myself. That said, I think it may now be worth considering whether we can start moving these core issuer types outside of the core cert-manager codebase so that individual issuer authors can evolve their code without potentially effecting other issuer implementations 😄

wallrj commented:
In https://github.com/jetstack/cert-manager/pull/3412#pullrequestreview-515527111 @munnerz wrote:
> I voiced this on the stand up call that I don't think cert-manager should be persisting the refresh token and relying on it as the means to authenticate. Given this is specific to the Venafi issuer, I don't want to 'hold' this back if you feel it is in fact okay.

James,

I have read a lot since that meeting, chatted with people at Venafi and thought about it some more.
It's time I summarized what I've learned, and time closed a few dozen oauth2 related browser tabs!

Here are my responses to your points:

> The problem areas I see here:

## 1

> 1) writing back into the same user-provided Secret resource: this is something that has caused us a load of problems in that past with the Certificate resources `spec.secretName` - we conflated the 'input' (private key) and the output (certificate) into a single resource. In a similar manner, we are conflating the input, and the output (the refresh token)

I agree, but note that this is a WIP PR.
This is the result of me hacking on the existing code to see what needs to change.
I would like to get the code working with immutable secrets and have it write the refreshed access-token to a new immutable Secret.
Deleting the previous secret once it has been stored.

## 2

> 2) refresh tokens are not typically persisted this way. In a usual oauth flow, some other credential is exchanged for an access token and refresh token. This allows for the client to re-authenticate in the event of the refresh token expiring. Making the refresh token valid for extremely long periods of time to avoid this undermines the integrity of oauth2 altogether, and you'd be better to avoid all this complexity and simply have a long lived access token (or username/password pair). They are functionally equivalent, and mixing oauth into this is introducing unnecessary complexity.

### "refresh tokens are not typically persisted this way"

I disagree.
`gcloud`, for example, writes access-token and refresh-token to disk and uses the long-lived refresh-token to update the short lived access-token.
You can see them if you run `sqlite3 ~/.config/gcloud/credentials.db 'select value from credentials'`.
The refresh-token in this case is returned as a result of requesting `offline_access`

It is not always clearly documented, but here is a an example from Google's Youtube API docs, expecting the refresh-token to be persisted.
[Refreshing an access token (offline access)](https://developers.google.com/youtube/v3/live/guides/auth/server-side-web-apps#OAuth2_Refreshing_a_Token) says:

> Note that there are limits on the number of refresh tokens that will be issued; one limit per client/user combination, and another per user across all clients. You should save refresh tokens in long-term storage and continue to use them as long as they remain valid. If your application requests too many refresh tokens, it may run into these limits, in which case older refresh tokens will stop working

And [Using OAuth 2.0 to Access Google APIs](https://developers.google.com/identity/protocols/oauth2#5.-refresh-the-access-token,-if-necessary) which says:

> Note: Save refresh tokens in secure long-term storage and continue to use them as long as they remain valid. Limits apply to the number of refresh tokens that are issued per client-user combination, and per user across all clients, and these limits are different. If your application requests enough refresh tokens to go over one of the limits, older refresh tokens stop working.

There's more discussion of this in the SO document [Do Google refresh tokens expire?](https://stackoverflow.com/questions/8953983/do-google-refresh-tokens-expire).
And in the  [OIDC Spec for Offline Access](https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess).

### "In a usual oauth flow, some other credential is exchanged for an access token and refresh token":

We will ask users to do this in a separate step [using the `vcert getcred` command and entering the username and password, or a p12 encoded client certificate](https://github.com/cert-manager/website/pull/346/files#diff-96d90f67ba6a29316334e94fa90bf0af82ea5c0c334715f3159db0a8e1a78d0aR169-R187)

### "some other credential is exchanged for an access token and refresh token. This allows for the client to re-authenticate in the event of the refresh token expiring":

Yeah, but then the user credentials need to be stored in the Kubernetes cluster which is one of the things we're trying to avoid.
Why?
Because if a TPP username-password or a client-cert is leaked it allows an attacker to do everything that that user is permitted.
For example, it would allow the holder to access the TPP web admin pages,
and from there perform all the operations granted to that user in the webUI.
For example they could change the password, to create and delete teams, join, and remove members from teams.
Or create their own access-tokens and refresh-tokens with applications and API scopes of their choice.
including some or all of manage SSH keys, Certificates, SSH keys, Configuration etc.

And how do you revoke a leaked user account credential?
You'd need to change the password or disable the account in TPP.
And if other cert-managers or other integrated applications are also using that account, you would need to go and reconfigure those too.

We could suggest cert-manager users should create separate least-privilege accounts for each cert-manager Venafi Issuer.
And I think you had this in mind when you drew the comparison with Google Service Accounts.
TPP has user accounts and teams.
But it does not have an equivalent to Google Service Accounts.

Google's [Understanding service accounts](https://cloud.google.com/iam/docs/understanding-service-accounts#managing_service_accounts) document says:

> Note that service accounts can be thought of as both a resource and as an identity.
> When thinking of the service account as an identity, you can grant a role to a service account, allowing it to access a resource (such as a project).
> When thinking of a service account as a resource, you can grant roles to other users to access or manage that service account.

So unlike a TPP user account:
 * a service-account can be scoped to particular API endpoints and API methods.
 * a service-account can be managed by another user account.
 * and a [service account credentials can be easily and programatically rotated](https://cloud.google.com/iam/docs/creating-managing-service-account-keys#deleting_service_account_keys), which becomes more important the more places they are stored.
 * [...plus other differences](https://cloud.google.com/iam/docs/service-accounts#differences_between_a_service_account_and_a_user_account)

And even if TPP did have all these service-account features, I'm still not sure it would be appropriate to use them.
Google's [Using OAuth 2.0 for Server to Server Applications](https://developers.google.com/identity/protocols/oauth2/service-account) document says this:

> Typically, an application uses a service account when the application uses Google APIs to work with its own data rather than a user's data.
> For example, an application that uses Google Cloud Datastore for data persistence would use a service account to authenticate its calls to the Google Cloud Datastore API.

Well I'd argue that a service-account is not appropriate here, since we are trying to provide cert-manager with permission to manage another users data, not its own.

Kubernetes User A is configuring cert-manager Issuer B to have access to their data on TPP service C.
This sounds a lot like [3-legged Oauth2](https://stackoverflow.com/a/13384100/919436) to me.

### "Making the refresh token valid for extremely long periods of time to avoid this undermines the integrity of oauth2 altogether"

I disagree.

1. First, I want to make clear that in TPP, the refresh-tokens are *one-time-use* and *valid for 365 days* by default.
TPP implements the [Refresh Token Rotation mechanism described in the Oauth2 RFC](https://tools.ietf.org/html/rfc6749#section-10.4).
This is a [protection against refresh-token abuse](https://mailarchive.ietf.org/arch/msg/oauth/e5iV5x-r8U1iLQPX_zOvHtal2rY/).

The problem is then how to reliably persist the refresh-token each time it is rotated, which I've written more below.

2. Second, I have shown above that Google DO hand out *long-lived* and *re-usable* oauth2 refresh-tokens and advise users to store them.

Storing a long-lived re-usable refresh-token is an improvement over storing the users credentials or storing a long-lived access-token.
The refresh-token can only be used to get a new short-lived access-token, and one which has the same limited scope as the previous access-token.
The refresh-token, if leaked, can be revoked without affecting any of the other applications which are using tokens associated with the same user account.

In contrast, a long-lived ["self-encoded" access-token can not so easily be revoked](https://www.oauth.com/oauth2-servers/listing-authorizations/revoking-access/#self-encoded),
because the resource API server will accept it by virtue of it having the correct signed checksum.
Some oauth2 implementations have a mechanism for revoking access-tokens but not all.

In any case, I think TPP does support revoking access-tokens and refresh-tokens.
The API document [GET Revoke/Token](https://docs.venafi.com/Docs/current/TopNav/Content/SDK/AuthSDK/r-SDKa-GET-Revoke-Token.php) says:
> Revokes the caller's Token Auth grant and blocks the ability to make Web SDK calls. This method also invalidates the refresh token.

### "you'd be better to avoid all this complexity and simply have a long lived access token"

I agree that long-lived access token should be a configuration option,
and I implemented that in the PR [Add support for long lived TPP access-token credentials](https://github.com/jetstack/cert-manager/pull/3379).
The TPP web UI has a checkbox where you can disable the refresh-token feature, so this is important for TPP those users.

### "simply have a long lived access token (or username/password pair). They are functionally equivalent, and mixing oauth into this is introducing unnecessary complexity."

I disagree.
As described above, a TPP a username-password gives wider access than an access-token.
It is harder to revoke than an access-token.
And there is no API for changing a password, so it would be difficult to rotate the password,
which if it was stored in the Kubernetes cluster, you would want to do regularly.

# 3
> 3) the venafi client stuff here was intended to be a read-only cache, where the cache is populated by a controller that watches for venafi-typed issuer resources. This creates a clear separation between what constructs Venafi clients, and what utilises them. This change now puts a *write* operation onto the *read* path. If you are going to do this, I'd strongly recommend adding the logic for persisting and rotating refresh tokens into a distinct controller and keeping the *read* path as simple as possible (i.e. looking up a 'venafi client' in a map of issuer->client structure)

That's a good idea.
So how about if the Venafi Issuer Setup function was responsible for refreshing the token.
And a separate controller for triggering the Issuer and ClusterIssuer controllers a number of minutes before the access-token is about to expire.
Perhaps by adding / updating a condition on the Issuer.Status of Venafi Issuer resources, thus triggering the Issuer controller.

I will investigate that.

> Overall, I'm quite strongly against this change (as we've already discussed). I think it reduces reliability, increases complexity, and has a net *negative* on the security of the system.

Hopefully I've addressed some of your concerns.

> Given this is ultimately a vendor-specific change however, I wouldn't want to put an executive 'hold' on it myself. That said, I think it may now be worth considering whether we can start moving these core issuer types outside of the core cert-manager codebase so that individual issuer authors can evolve their code without potentially effecting other issuer implementations 😄

I agree.
I'll talk to the others about breaking out the Venafi Issuer and perhaps the Vault Issuer too.

wallrj commented:
I added a bit more background context to a README file (for now)...but it's a work in progress.

meyskens commented:
/test pull-cert-manager-e2e-v1-19

maelvls commented:
I just came across your comment above, what a comment!! Jaw-dropping 🤯
